### PR TITLE
Remove Shell.GenericContainer

### PR DESCRIPTION
**Pull request by GeorgesStavracas**
This is it; this is the final patchset. The details of each change is described in each commit's message, but the basic pattern around here was:

 * Turn the class into an St.Widget (or St.Bin or other suitable classes) subclass;
 * Tentatively use Clutter's built-in features such as layout managers, etc, to avoid overriding vfuncs;
 * If not possible, override size negotiation vfuncs;
 * Remove Shell.GenericContainer;

Downstream commits:
 * https://github.com/endlessm/gnome-shell/pull/298/commits/19a53fbdeb802a1c9de0da16544f77d2200dacd0
 * https://github.com/endlessm/gnome-shell/pull/298/commits/13076f6cb158ec91dc999e42b3db44e8bf03aa3b
 * https://github.com/endlessm/gnome-shell/pull/298/commits/a1d6c29b15fc31958dd94701767d2862b255ef5c
 * https://github.com/endlessm/gnome-shell/pull/298/commits/398bd6de3cab427f316669fd9ef24610ab875a22

https://phabricator.endlessm.com/T23147

### CODE REVIEW

File: js/ui/layout.js [hunk 1]
```diff
@@ -84,23 +84,12 @@ var LayoutManager = new Lang.Class({
...
-                            let width = global.stage.width;
-                            [alloc.min_size, alloc.natural_size] = [width, width];
-                        });
-        this.uiGroup.connect('get-preferred-height',
-                        function(actor, forWidth, alloc) {
-                            let height = global.stage.height;
-                            [alloc.min_size, alloc.natural_size] = [height, height];
-                        });
+        this.uiGroup = new St.Widget({ name: 'uiGroup' });
+        this.uiGroup.set_flags(Clutter.ActorFlags.NO_LAYOUT);
```
Comment (cosimoc) [inline js/ui/layout.js]: Does this not belong to a separate commit?

File: js/gdm/loginDialog.js [hunk 2]
```diff
@@ -921,11 +928,12 @@ var LoginDialog = new Lang.Class({
         this._showPrompt();
     },
 
-    _loginScreenSessionActivated: function() {
-        if (this.actor.opacity == 255 && this._authPrompt.verificationStatus == AuthPrompt.AuthPromptStatus.NOT_VERIFYING)
+
```
Comment (cosimoc) [inline js/gdm/loginDialog.js]: Remove extra empty line

File: tests/interactive/scroll-view-sizing.js [hunk 3]
```diff
@@ -32,26 +32,22 @@ const BOX_WIDTHS = [
...
 	                             style: 'border: 1px solid #444444; background: #00aa44' })
-	    this.actor.add_actor(child);
+	    this.add_actor(child);
 	}
     },
 
-    _getPreferredWidth: function (actor, forHeight, alloc) {
-        let children = this.actor.get_children();
+    vfunc_get_preferred_width(forHeight) {
+        let children = this.get_children();
```
Comment (cosimoc) [inline tests/interactive/scroll-view-sizing.js]: Indentation looks off in a few places in this patch

File: tests/interactive/scroll-view-sizing.js [hunk 4]
```diff
@@ -184,27 +177,28 @@ SizingIllustrator.prototype = {
...
+    vfunc_get_preferred_height(forWidth) {
+        let children = this.get_children();
 	for (let i = 0; i < children.length; i++) {
 	    let child = children[i];
 	    if (child == this.child) {
 		[this.minHeight, this.naturalHeight] = child.get_preferred_height(this.width);
 	    } else {
-		let [minWidth, naturalWidth] = child.get_preferred_width(-1);
-		child.get_preferred_height(naturalWidth);
+		let [minWidth, naturalWidth] = child.get_preferred_height(naturalWidth);
```
Comment (cosimoc) [inline tests/interactive/scroll-view-sizing.js]: The passed in `naturalWidth` seems undefined here. Also it does not make sense that the assigned variables are called `minWidth`/`naturalWidth` but we're querying the preferred height

File: js/ui/workspaceSwitcherPopup.js [hunk 5]
```diff
@@ -159,17 +175,14 @@ var WorkspaceSwitcherPopup = new Lang.Class({
...
             Mainloop.source_remove(this._timeoutId);
         this._timeoutId = 0;
 
         for (let i = 0; i < this._globalSignals.length; i++)
             global.screen.disconnect(this._globalSignals[i]);
 
-        this.actor.destroy();
-
-        this.emit('destroy');
+        this._workspaceManagerSignals = [];
```
Comment (cosimoc) [inline js/ui/workspaceSwitcherPopup.js]: Should this not be `this._globalSignals`?

File: js/ui/boxpointer.js [hunk 6]
```diff
@@ -80,6 +80,36 @@ var BoxPointer = new Lang.Class({
         }
     },
 
+    // BoxPointer.show() and BoxPointer.hide() are here for only compatibility
+    // purposes, and will be removed in 3.32.
+    show(animate, onComplete) {
```
Comment (cosimoc) [inline js/ui/boxpointer.js]: We don't really need this commit though, right? Since we don't care about extensions.

File: js/ui/panelMenu.js [hunk 7]
```diff
@@ -206,6 +210,8 @@ var Button = new Lang.Class({
     },
 
     _onDestroy() {
+        this.parent();
```
Comment (cosimoc) [inline js/ui/panelMenu.js]: Part of this patch needs to be squashed with an earlier one (I had a comment about that)

File: js/ui/status/keyboard.js [hunk 8]
```diff
@@ -829,6 +829,48 @@ function getInputSourceManager() {
     return _inputSourceManager;
 }
 
+var InputSourceIndicatorContainer = new Lang.Class({
+    Name: 'InputSourceIndicatorContainer',
+    Extends: St.Widget,
+
+    _init() {
```
Comment (cosimoc) [inline js/ui/status/keyboard.js]: This could be removed

File: js/ui/appIconBar.js [hunk 9]
```diff
@@ -466,27 +472,27 @@ const AppIconButton = new Lang.Class({
 
         // Calculate location of the label only if we're not tweening as the
         // values will be inaccurate
-        if (!Tweener.isTweening(this.actor)) {
-            let iconMidpoint = this.actor.get_transformed_position()[0] + this.actor.width / 2;
+        if (!Tweener.isTweening(this)) {
+            let iconMidpoint = this.get_transformed_position()[0] + this.width / 2;
```
Comment (cosimoc) [inline js/ui/appIconBar.js]: Would be nice to save the output of `this.get_transformed_position()` into local variables, since it's used below too.

File: js/ui/layout.js (hunk 1 above)
Comment (GeorgesStavracas) [reply js/ui/layout.js]: Ouch. This was a performance touch from a commit upstream, and it sneaked there while i was rebasing the patchset.

File: tests/interactive/scroll-view-sizing.js (hunk 3 above)
Comment (GeorgesStavracas) [reply tests/interactive/scroll-view-sizing.js]: Fixed a few tabs that weren't removed in the lines I touched though.

**Comment by GeorgesStavracas** [comment]
Thanks @cosimoc. Your points were all addressed now.

Notice that the problematic `_onDestroyed` signal callback was fixed, but it's probably not what you expected; the correct order (which is not what the patchset does) is:

 1. Initially only `PanelMenu.Button` connects to it;
 2. Signal moves to parent class `PanelMenu.ButtonBox`

**Comment by cosimoc** [comment]
Let’s try this! Looks good to me now.

**Comment by cosimoc** [comment]
@GeorgesStavracas please also update the upstream MR with any change from the last round of review, if applicable, even if getting it merged upstream is out of scope for 3.5.0.

**Comment by GeorgesStavracas** [comment]
Sure, I'm already working on that
